### PR TITLE
[richiehx] iP

### PR DESCRIPTION
**Pull request by richiehx**
# Nano the Chatbot
> "If ur asking, u want it" - Anonymous

Nano frees your mind of having to remember things you need to do. It's,
- Text-based
- Easy to learn
- Killing me

All you need to do is,
1. download it from [here](https://github.com/richiehx/ip/releases/tag/A-Jar).
2. double-click it.
3. add your tasks.
4. Try not to cry. :droplet:
5. ???
6. Profit :+1:

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---
If you Java programmer, you can use it to practice Java too. Here's the main method:
```
public static void main(String[] args) {
        new Duke().run();
    }
```

### CODE REVIEW

File: src/main/java/duke/command/Command.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+public enum Command {
+    BYE("bye"),
+    LIST("list"),
+    MARK("mark"),
+    UNMARK("unmark"),
+    DELETE("delete"),
+    TODO("todo"),
+    DEADLINE("deadline"),
+    EVENT("event"),
+    FIND("find");
```
Comment (LWZ19) [inline src/main/java/duke/command/Command.java]: Nice use of enums!! 😄

File: src/main/java/duke/Duke.java [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+            try {
+                switch (command) {
+                    case BYE:
+                        ui.printBye();
+                        return;
+                    case LIST:
+                        ui.printOutput(tasks.formatList());
+                        break;
+                    case MARK:
```
Comment (LWZ19) [inline src/main/java/duke/Duke.java]: Perhaps you could include a `//fallthrough` comment, for example
```suggestion
                    case MARK: //fallthrough
```

File: src/main/java/duke/Duke.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+                        break;
+                    case FIND:
+                        String findQuery = parseLine.parseFindQuery();
+                        ui.printOutput(tasks.findTasks(findQuery).formatList());
+                        break;
+                    default:
+                        ui.printOutput("I don't understand what you're saying.");
+                        break;
+                }
+            } catch (DukeException e) {
```
Comment (LWZ19) [inline src/main/java/duke/Duke.java]: Good job for using switch statements, which makes your code clean and easy to understand!
But, perhaps you could abstract out the handling of the different commands to another class?

File: src/main/java/duke/Duke.java [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    /** The storage that is used to save and load the user's tasks. */
+    private static Storage storage;
+
+    /** The task list that is used to store the user's tasks. */
+    private TaskList tasks;
+
+    /** The UI to interact with the User. */
+    private Ui ui;
+
+    /** The file path to store the User's Task Data. */
```
Comment (LWZ19) [inline src/main/java/duke/Duke.java]: I like how you took the effect of writing JavaDoc comments for each attribute and method! :thumbsup:

File: src/main/java/duke/command/Command.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+    /**
+     * The String value of the command.
+     */
+    private String value;
```
Comment (LWZ19) [inline src/main/java/duke/command/Command.java:21]: Perhaps a more intuitive variable name here? So that it is consistent with the getter method.

File: src/main/java/duke/parser/Parser.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+            throw new DukeException("Hey, the Event given is Invalid!" +
+                    " Make sure that you follow this format:\n" +
+                    Ui.INDENTATION + " 'eventDescription /from startTime /to endTime'");
+        }
+        String start = startEnd[0];
+        String end = startEnd[1];
+
+        return new String[]{desc, start, end};
+    }
+
```
Comment (LWZ19) [inline src/main/java/duke/parser/Parser.java:169]: Great work, It's nice to see you have different parse methods for different commands!

File: src/main/java/duke/parser/Parser.java [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+                    " Make sure that you follow this format:\n" +
+                    Ui.INDENTATION + " 'eventDescription /from startTime /to endTime'");
+        }
+        String desc = arr[0];
+
+        String[] startEnd = arr[1].split(" /to ");
+        if (startEnd.length != 2) {
+            throw new DukeException("Hey, the Event given is Invalid!" +
+                    " Make sure that you follow this format:\n" +
+                    Ui.INDENTATION + " 'eventDescription /from startTime /to endTime'");
```
Comment (LWZ19) [inline src/main/java/duke/parser/Parser.java]: Perhaps you could abstract out different exceptions tailored to different problems? So that you can reduce the amount of reputation in the code.

File: src/test/java/duke/parser/ParserTest.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
+package duke.parser;
+
+import duke.Duke;
+import duke.exception.DukeException;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
```
Comment (LWZ19) [inline src/test/java/duke/parser/ParserTest.java]: It would be better to import everything you need, rather than using the wildcard('`*`')

File: src/main/java/duke/tasks/TaskList.java [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+
+import java.lang.reflect.Array;
+import java.util.ArrayList;
+
+/**
+ * Represents a Task List to store User's Tasks.
+ */
+public class TaskList {
+    /** The Indentation Level to format text. */
+    private static final String INDENTATION = Ui.INDENTATION;
```
Comment (LWZ19) [inline src/main/java/duke/tasks/TaskList.java]: Any reason why you did it this way instead of using `Ui.INDENTATION` directly?

File: src/main/java/duke/tasks/TaskList.java [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+            String buffer = "";
+
+            if (index > taskList.size() - 1 || index < 0) {
+                throw new DukeException("Invalid duke.tasks.Task Index provided!");
+            }
+            Task selectedTask = taskList.get(index);
+            if (status.equals("mark")) {
+                selectedTask.markTask();
+                buffer = "Nice! I've marked the task as done:\n   " +
+                        INDENTATION + selectedTask;
```
Comment (LWZ19) [inline src/main/java/duke/tasks/TaskList.java]: Perhaps you could add the line break before the `+` operator.
```suggestion
                buffer = "Nice! I've marked the task as done:\n   "
                         + INDENTATION + selectedTask;
```

File: src/main/java/duke/ui/Ui.java [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+    /**
+     * Prints the formatted output of the Chatbot.
+     *
+     * @param rawOutput The raw output to be formatted and printed.
+     */
+    public void printOutput(String rawOutput) {
+        System.out.println(formatOutput(rawOutput));
+    }
+}
```
Comment (LWZ19) [inline src/main/java/duke/ui/Ui.java:72]: Nice work on the text UI! :blush: Perhaps you could include more features in the Ui, for example, Ui can be used to handle most of the display interactions with the user.

File: src/test/java/duke/parser/ParserTest.java [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+
+public class ParserTest {
```
Comment (LWZ19) [inline src/test/java/duke/parser/ParserTest.java:12]: Nice to see you have written some unit tests!

**Review by LWZ19 — COMMENTED** [review]
LGTM! Some suggestions are your UI class could be used to handle more user interactions and just some small nits on the line break before an operator.

Overall, you have done an excellent job and is on par with code qualities and standard. Very intuitive naming for methods and variables. Keep it up! 🎉

File: src/main/java/duke/Duke.java [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+
+    /**
+     * The main method is used to run the program.
+     *
+     * @param args The command line arguments.
+     */
+    public static void main(String[] args) {
+        new Duke().run();
+    }
+}
```
Comment (seanpzk) [inline src/main/java/duke/Duke.java:138]: I like how you packaged your different classes!

File: src/main/java/duke/tasks/Event.java [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+    /**
+     * Returns the String representation of the Event task.
+     *
+     * @return String representation of the Event task.
+     */
+    @Override
+    public String toString() {
+        return "[E]" + super.toString() + " (from: " + this.start + " to: " + this.end + ")";
+    }
+}
```
Comment (seanpzk) [inline src/main/java/duke/tasks/Event.java:105]: JavaDocs are very clear and concise :)

File: src/test/java/duke/parser/ParserTest.java [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+        Exception exception = assertThrows(DukeException.class, () -> {
+            parseLine.parseEventArguments();
+        });
+        String expected = "Hey, the Event given is Invalid!" +
+                " Make sure that you follow this format:\n"
+                + " 'eventDescription /from yyyy-mm-dd hh:mm /to yyyy-mm-dd hh:mm'";
+        assertEquals(expected, exception.getMessage());
+    }
+
+}
```
Comment (seanpzk) [inline src/test/java/duke/parser/ParserTest.java:165]: I like how the test names are clear and easy to know what you are trying to test for. Also good effort on implementing the different test cases to test both positive and negative test cases

File: src/main/java/duke/parser/Parser.java [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+            throw new DukeException("Hey, the Event given is Invalid!"
+                    + " Make sure that you follow this format:\n"
+                    + " 'eventDescription /from yyyy-mm-dd hh:mm /to yyyy-mm-dd hh:mm'");
+        }
+        String start = startEnd[0];
+        String end = startEnd[1];
+
+        return new String[]{desc, start, end};
+    }
+
```
Comment (seanpzk) [inline src/main/java/duke/parser/Parser.java:169]: Good job, the JavaDoc comments follow the coding standard closely and is clear in explaining the purposes of the functions and variables!
